traffic_ops/traffic_ops_golang/user/user.go

```diff
@@ -508,3 +578,483 @@ func (user *TOUser) InsertQuery() string {
+	var userV4 tc.UserV4
+	var err error
+	inf, userErr, sysErr, errCode := api.NewInfo(r, nil, nil)
+	if userErr != nil || sysErr != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, errCode, userErr, sysErr)
+		return
+	}
+	defer inf.Close()
+
+	tx := inf.Tx.Tx
+	if err := json.NewDecoder(r.Body).Decode(&userV4); err != nil {
+		api.HandleErr(w, r, tx, http.StatusBadRequest, err, nil)
+		return
+	}
+	if err := validateUserV4(userV4); err != nil {
+		api.HandleErr(w, r, tx, http.StatusBadRequest, err, nil)
+		return
+	}
+	if err := postValidateV40(userV4); err != nil {
+		api.HandleErr(w, r, tx, http.StatusBadRequest, err, nil)
+		return
+	}
+
+	toUser := TOUser{
+		APIInfoImpl: api.APIInfoImpl{ReqInfo: inf},
+	}
+	toUser.User = userV4.Downgrade()
+
+	authorized, err := toUser.IsTenantAuthorized(inf.User)
+	if err != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, nil, errors.New("checking tenant authorized: "+err.Error()))
+		return
+	}
+	if !authorized {
+		api.HandleErr(w, r, inf.Tx.Tx, http.StatusForbidden, errors.New("not authorized on this tenant"), nil)
+		return
+	}
+
+	// Convert password to SCRYPT
+	*userV4.LocalPassword, err = auth.DerivePassword(*userV4.LocalPassword)
+	if err != nil {
+		api.HandleErr(w, r, tx, http.StatusBadRequest, err, nil)
+		return
+	}
+
+	var resultRows *sqlx.Rows
+	_, ok, err := dbhelpers.GetRoleIDFromName(tx, userV4.Role)
+	if err != nil {
+		api.HandleErr(w, r, tx, http.StatusInternalServerError, nil, errors.New("no ID exists for the supplied role name"))
+		return
+	} else if !ok {
+		api.HandleErr(w, r, tx, http.StatusNotFound, errors.New("role not found"), nil)
+		return
+	}
+	resultRows, err = inf.Tx.NamedQuery(InsertQueryV40(), userV4)
+	if err != nil {
+		userErr, sysErr, statusCode := api.ParseDBError(err)
+		api.HandleErr(w, r, tx, statusCode, userErr, sysErr)
+		return
+	}
+	defer resultRows.Close()
+
+	var id int
+	var lastUpdated time.Time
+	var tenant string
+	var rolename string
+	var changeLogMsg string
+
+	rowsAffected := 0
+	for resultRows.Next() {
+		rowsAffected++
+		if err = resultRows.Scan(&id, &lastUpdated, &tenant, &rolename); err != nil {
+			api.HandleErr(w, r, tx, http.StatusInternalServerError, nil, fmt.Errorf("could not scan after insert: %s\n)", err))
+			return
+		}
+	}
+
+	if rowsAffected == 0 {
+		api.HandleErr(w, r, tx, http.StatusInternalServerError, nil, fmt.Errorf("no userV4 was inserted, nothing was returned"))
+		return
+	} else if rowsAffected > 1 {
+		api.HandleErr(w, r, tx, http.StatusInternalServerError, nil, fmt.Errorf("too many rows affected from userV4 insert"))
+		return
+	}
+
+	userV4.ID = &id
+	userV4.LastUpdated = lastUpdated
+	userV4.Tenant = &tenant
+	userV4.Role = rolename
+	userV4.LocalPassword = nil
+
+	userResponse := tc.UserResponseV4{
+		Response: userV4,
+		Alerts:   tc.CreateAlerts(tc.SuccessLevel, "user was created."),
+	}
+	api.WriteAlertsObj(w, r, http.StatusOK, userResponse.Alerts, userResponse.Response)
```

Comment:
status code for object creation should be `200 OK`, ideally with an HTTP `Location` header set to a URI that can be used to retrieve a representation of the created object.

Reply:
Do you mean a `201 Created`?